### PR TITLE
[SPARK-7222][ML] Added mathematical derivation in comment and compressed the model, removed the correction terms in LinearRegression with ElasticNet

### PR DESCRIPTION
**Pull request by dbtsai**
Added detailed mathematical derivation of how scaling and LeastSquaresAggregator work. Refactored the code so the model is compressed based on the storage. We may try compression based on the prediction time. 

Also, I found that diffSum will be always zero mathematically, so no corrections are required.

### CODE REVIEW

File: mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala [hunk 1]
```diff
@@ -110,59 +110,70 @@ class LinearRegression extends Regressor[Vector, LinearRegression, LinearRegress
...
-
-    val costFun = new LeastSquaresCostFun(instances, yStd, yMean,
-      featuresStd, featuresMean, effectiveL2RegParam)
-
-    val optimizer = if (paramMap(elasticNetParam) == 0.0 || effectiveRegParam == 0.0) {
-      new BreezeLBFGS[BDV[Double]](paramMap(maxIter), 10, paramMap(tol))
-    } else {
-      new BreezeOWLQN[Int, BDV[Double]](paramMap(maxIter), 10, effectiveL1RegParam, paramMap(tol))
-    }
+    // If the yStd is zero, then the intercept is yStd with zero weights.
```
Comment (viirya) [inline mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala]: ...the intercept is yMean...

Comment (dbtsai) [inline mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala]: thanks.

File: mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala [hunk 2]
```diff
@@ -110,59 +110,70 @@ class LinearRegression extends Regressor[Vector, LinearRegression, LinearRegress
...
-      featuresStd, featuresMean, effectiveL2RegParam)
-
-    val optimizer = if (paramMap(elasticNetParam) == 0.0 || effectiveRegParam == 0.0) {
-      new BreezeLBFGS[BDV[Double]](paramMap(maxIter), 10, paramMap(tol))
-    } else {
-      new BreezeOWLQN[Int, BDV[Double]](paramMap(maxIter), 10, effectiveL1RegParam, paramMap(tol))
-    }
+    // If the yStd is zero, then the intercept is yMean with zero weights;
+    // as a result, training is not needed.
+    val model = if (yStd != 0.0) {
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala]: Is it really rare? If `yStd` is `0.0` and then the optimal model would be empty with intercept `yMean`. In this case, a warning would be proper. Having this giant `if ... else` block making the code hard to read.

~~~scala
if (yStd == 0.0) {
  logWarning(...)
  if (handlePersistence) ...
  return new LinearRegressionModel(...)
}

// actual implementation
~~~

Comment (dbtsai) [inline mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala]: fair!

File: mllib/src/main/scala/org/apache/spark/mllib/util/LinearDataGenerator.scala [hunk 3]
```diff
@@ -103,17 +103,16 @@ object LinearDataGenerator {
...
-      var i = 0
-      while (i < vectorArray.size) {
-        vectorArray(i) = (vectorArray(i) - 0.5) * math.sqrt(12.0 * xVariance(i)) + xMean(i)
-        i += 1
-      }
-    })
+      Array.fill[Double](weights.length)(rnd.nextDouble()))
+
+    x.foreach {
+      case v =>
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/mllib/util/LinearDataGenerator.scala]: minor: `x.foreach { v =>`

**Comment by mengxr** [comment]
LGTM. Merged into master. Thanks! @dbtsai Please address the comment in a separate PR.

**Pull request closed without merging**